lambda/value_rec_compiler.ml

```diff
@@ -0,0 +1,792 @@
+    in
+    begin match sw_consts_res, sw_blocks_res, sw_failaction_res with
+    | Unreachable, Unreachable, (None | Some Unreachable) -> Unreachable
+    | Reachable (lfun, sw_consts), Unreachable, (None | Some Unreachable) ->
+      Reachable (lfun, Lswitch (arg, { sw with sw_consts }, loc))
+    | Unreachable, Reachable (lfun, sw_blocks), (None | Some Unreachable) ->
+      Reachable (lfun, Lswitch (arg, { sw with sw_blocks }, loc))
+    | Unreachable, Unreachable, Some (Reachable (lfun, failaction)) ->
+      let switch =
+        Lswitch (arg, { sw with sw_failaction = Some failaction }, loc)
+      in
+      Reachable (lfun, switch)
+    | Reachable _, Reachable _, _ | Reachable _, _, Some (Reachable _)
+    | _, Reachable _, Some (Reachable _) ->
+      Misc.fatal_error "letrec: multiple functions"
+    end
+  | Lstringswitch (arg, arms, failaction, loc) ->
+    let arms_res = rebuild_arms block_var local_idents arms in
+    let failaction_res =
+      Option.map (split_static_function block_var local_idents) failaction
+    in
+    begin match arms_res, failaction_res with
+    | Unreachable, (None | Some Unreachable) -> Unreachable
+    | Reachable (lfun, arms), (None | Some Unreachable) ->
+      Reachable (lfun, Lstringswitch (arg, arms, failaction, loc))
+    | Unreachable, Some (Reachable (lfun, failaction)) ->
+      Reachable (lfun, Lstringswitch (arg, arms, Some failaction, loc))
+    | Reachable _, Some (Reachable _) ->
+      Misc.fatal_error "letrec: multiple functions"
+    end
+  | Lstaticcatch (body, (nfail, params), handler) ->
+    let body_res = split_static_function block_var local_idents body in
+    let handler_res =
+      let local_idents =
+        List.fold_left (fun vars (var, _) -> Ident.Set.add var vars)
+          local_idents params
+      in
+      split_static_function block_var local_idents handler
+    in
+    begin match body_res, handler_res with
+    | Unreachable, Unreachable -> Unreachable
+    | Reachable (lfun, body), Unreachable ->
+      Reachable (lfun, Lstaticcatch (body, (nfail, params), handler))
+    | Unreachable, Reachable (lfun, handler) ->
+      Reachable (lfun, Lstaticcatch (body, (nfail, params), handler))
+    | Reachable _, Reachable _ ->
+      Misc.fatal_error "letrec: multiple functions"
+    end
+  | Ltrywith (body, exn_var, handler) ->
+    let body_res = split_static_function block_var local_idents body in
+    let handler_res =
+      split_static_function block_var
+        (Ident.Set.add exn_var local_idents) handler
+    in
+    begin match body_res, handler_res with
+    | Unreachable, Unreachable -> Unreachable
+    | Reachable (lfun, body), Unreachable ->
+      Reachable (lfun, Ltrywith (body, exn_var, handler))
+    | Unreachable, Reachable (lfun, handler) ->
+      Reachable (lfun, Ltrywith (body, exn_var, handler))
+    | Reachable _, Reachable _ ->
+      Misc.fatal_error "letrec: multiple functions"
```

Comment:
Is this the result of calling `split_static_function` on a non-static function? Then maybe we want to produce a similar error message as below ("letrec binding is not a static function")?

Reply:
Like all occurrences of `Misc.fatal_error`, it is supposed to be unreachable unless there is a bug in the compiler.
Having a different message helps a little bit in the unlikely event that we actually have to debug this code; here I would know if I get this message that I got a term where two different tails are functions, while the "not a static function" message would tell me that I got a term whose size was computed as `Function` but I couldn't find the function.
This code should be unreachable because `Value_rec_check` wouldn't allow it, and `join_sizes` also has a small comment mentioning this particular issue (that we can't join two `Function` sizes with the current algorithm even if `Value_rec_check` allowed it).